aea/helpers/dialogue/base.py

```diff
@@ -182,70 +289,169 @@ def outgoing_extend(self, message: "Message") -> None:
+        """
+        Extend the list of incoming/outgoing messages with 'message', if 'message' is valid
+
+        :param message: a message to be added
+        :return: True if message successfully added, false otherwise
+        """
+        if self.is_valid_next_message(message):
+            if message.is_incoming:
+                self.incoming_extend(message)
+            else:
+                self.outgoing_extend(message)
+            return True
+        else:
+            return False
+
+    @abstractmethod
+    def is_valid_next_message(self, message: "Message") -> bool:
+        """
+        Check whether 'message' is a valid next message in this dialogue.
+
+        :param message: the message to be validated
+        :return: True if yes, False otherwise.
+        """
+
+    @staticmethod
+    @abstractmethod
+    def role_from_first_message(message: Message) -> "Role":
+        """
+        Infer the role of the agent from an incoming/outgoing first message
+
+        :param message: an incoming/outgoing first message
+        :return: The role of the agent
+        """
+
+    def __str__(self) -> str:
+        """
+        Get the string representation.
+
+        :return: The string representation of the dialogue
+        """
+        rep = "Dialogue Label: " + str(self.dialogue_label) + "\n"
+
+        if self.is_self_initiated:
+            all_messages = [
+                msg
+                for msg in itertools.chain(
+                    *itertools.zip_longest(
+                        self._outgoing_messages, self._incoming_messages
+                    )
+                )
+                if msg is not None
+            ]
+        else:
+            all_messages = [
+                msg
+                for msg in itertools.chain(
+                    *itertools.zip_longest(
+                        self._incoming_messages, self._outgoing_messages
+                    )
+                )
+                if msg is not None
+            ]
+
+        for msg in all_messages:
+            rep += str(msg.get("performative")) + "( )\n"
+
+        rep = rep[:-1]
+        return rep
+
 
 class Dialogues:
-    """The dialogues class keeps track of all dialogues."""
+    """The dialogues class keeps track of all dialogues for an agent."""
 
-    def __init__(self) -> None:
+    def __init__(self, agent_address: Address) -> None:
         """
         Initialize dialogues.
 
+        :param agent_address: the address of the agent for whom dialogues are maintained
         :return: None
         """
         self._dialogues = {}  # type: Dict[DialogueLabel, Dialogue]
+        self._agent_address = agent_address
         self._dialogue_nonce = 0
 
     @property
     def dialogues(self) -> Dict[DialogueLabel, Dialogue]:
-        """Get dictionary of dialogues in which the agent is engaged in."""
+        """Get dictionary of dialogues in which the agent engages."""
         return self._dialogues
 
```

Comment:
We need to maintain all the historic APIs, although you can remove the decorator `@abstractmethod` so we no longer require them to be implemented in the derived class

Reply:
Done